x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,376 @@
+                Optional<String> maybeError = nodeHasCapacity(currentState, params, node);
+                if (maybeError.isPresent()) {
+                    try {
+                        builder.addFailedNode(params.getModelId(), node.getId(), maybeError.get());
+                    } catch (Exception e) {
+                        logger.warn(
+                            () -> new ParameterizedMessage(
+                                "[{}] unexpected failure when failing to allocate to node [{}] due to [{}]",
+                                params.getModelId(),
+                                node.getId(),
+                                maybeError.get()
+                            ),
+                            e
+                        );
+                        throw new ElasticsearchException(
+                            "Error when marking node [{}] allocation as failed with [{}]",
+                            e,
+                            node.getId(),
+                            maybeError.get()
+                        );
+                    }
+                } else {
+                    try {
+                        builder.addNode(params.getModelId(), node.getId());
+                    } catch (Exception e) {
+                        logger.warn(
+                            () -> new ParameterizedMessage(
+                                "[{}] unexpected failure when allocating to node [{}]",
+                                params.getModelId(),
+                                node.getId()
+                            ),
+                            e
+                        );
+                        throw new ElasticsearchException("Error when starting node [{}] allocation", e, node.getId());
+                    }
+                }
+            }
+        }
+        return update(currentState, builder);
+    }
+
+    static ClusterState updateModelRoutingTable(ClusterState currentState, UpdateTrainedModelAllocationStateAction.Request request) {
+        final String modelId = request.getModelId();
+        final String nodeId = request.getNodeId();
+        TrainedModelAllocationMetadata metadata = TrainedModelAllocationMetadata.metadata(currentState);
+        logger.trace(
+            () -> new ParameterizedMessage("[{}] [{}] current metadata before update {}", modelId, nodeId, Strings.toString(metadata))
+        );
+        final TrainedModelAllocation existingAllocation = metadata.getModelAllocation(modelId);
+
+        // If state is stopped, this indicates the node process is closed, remove the node from the allocation
+        if (request.getRoutingState().getState().equals(RoutingState.STOPPED)) {
+            if (existingAllocation == null || existingAllocation.routedToNode(nodeId) == false) {
+                return currentState;
+            }
+            return update(currentState, TrainedModelAllocationMetadata.builder(currentState).removeNode(modelId, nodeId));
+        }
```

Comment:
Maybe simply keep it as stopped until the node is removed or the allocation deleted? This way we don't get spurious allocation retries???